contracts/btc-finality/src/finality.rs

```diff
@@ -891,52 +899,62 @@ pub fn query_fps_by_total_active_sats(
-    let active_fps = get_power_table_at_height(deps.storage, env.block.height - 1)?;
-    // Short-circuit if there are no active finality providers
-    if active_fps.is_empty() {
-        return Ok(());
-    }
-    // Get the voting power of the active FPS
-    let total_voting_power = active_fps
-        .values()
-        .map(|power| *power as u128)
-        .sum::<u128>();
-    // Short-circuit if the total voting power is zero
-    if total_voting_power == 0 {
-        return Ok(());
-    }
-    // Get the rewards to distribute (bank balance of the finality contract, minus previously / already distributed rewards
-    // (pending to be sent to Babylon on an epoch boundary))
-    let total_pending_rewards = TOTAL_PENDING_REWARDS.load(deps.storage)?;
+/// Distributes rewards based on accumulated voting weights from the current reward interval.
+/// This approach uses pre-accumulated voting weights that are collected as FPs submit signatures,
+/// making the distribution O(n) instead of O(m*n) for much better gas efficiency.
+pub fn distribute_rewards_in_range(deps: &mut DepsMut, env: &Env) -> Result<(), ContractError> {
     let cfg = CONFIG.load(deps.storage)?;
-    let rewards_amount = deps
+
+    // Get current balance of the finality contract (total rewards to distribute)
+    let current_balance = deps
         .querier
         .query_balance(env.contract.address.clone(), cfg.denom)?
-        .amount
-        .saturating_sub(total_pending_rewards);
-    // Short-circuit if there are no rewards to distribute
-    if rewards_amount.is_zero() {
+        .amount;
+
+    if current_balance.is_zero() {
+        return Ok(()); // No rewards to distribute
+    }
+
+    // Read all accumulated voting weights from storage
+    let accumulated_weights: Vec<(String, u128)> = ACCUMULATED_VOTING_WEIGHTS
+        .range(deps.storage, None, None, cosmwasm_std::Order::Ascending)
+        .collect::<Result<Vec<_>, _>>()?;
+
+    if accumulated_weights.is_empty() {
+        return Ok(()); // No accumulated voting weights
+    }
+
+    // Calculate total accumulated voting weight
+    let total_accumulated_weight: u128 = accumulated_weights.iter().map(|(_, weight)| weight).sum();
+    if total_accumulated_weight == 0 {
         return Ok(());
     }
-    // Compute the rewards for each active FP
-    let mut accumulated_rewards = Uint128::zero();
-    for (fp_btc_pk_hex, power) in active_fps {
-        let reward = (rewards_amount.u128() * power as u128) / total_voting_power;
-        let reward = Uint128::from(reward);
-        // Update the rewards for this FP
-        REWARDS.update(deps.storage, &fp_btc_pk_hex, |r| {
-            Ok::<Uint128, ContractError>(r.unwrap_or_default() + reward)
-        })?;
-        // Compute the total rewards
-        accumulated_rewards += reward;
+
+    // Distribute rewards proportionally based on accumulated voting weights
+    for (fp_btc_pk_hex, accumulated_weight) in &accumulated_weights {
```

Comment:
`ACCUMULATED_VOTING_WEIGHTS.range` returns an iterator so we might be able to iterate over it directly, rather than reading those from DB and then iterating